closed/src/java.base/share/classes/openj9/internal/security/RestrictedSecurity.java

```diff
@@ -1207,64 +1697,71 @@ private static String parseProperty(String string) {
             return string;
         }
 
-        /**
-         * Check if the brackets are balanced.
-         *
-         * @param string input string for checking
-         * @return true if the brackets are balanced
-         */
-        private static boolean areBracketsBalanced(String string) {
-            Deque<Character> deque = new LinkedList<>();
-
-            for (char ch : string.toCharArray()) {
-                switch (ch) {
-                case '{':
-                    deque.addFirst('}');
-                    break;
-                case '[':
-                    deque.addFirst(']');
-                    break;
-                case '(':
-                    deque.addFirst(')');
-                    break;
-                case '}':
-                case ']':
-                case ')':
-                    if (deque.isEmpty() || (deque.removeFirst().charValue() != ch)) {
-                        return false;
+        private static void checkProviderFormat(String providerInfo, boolean update, boolean isFIPS140_2) {
+            Pattern p = Pattern.compile(
+                "^([A-Za-z0-9\\-\\.]+)"                             // provider name
+                + "\\s*(\\s[A-Za-z0-9\\$\\{\\}\\./]+)?"             // configuration file [optional]
+                + "\\s*(\\s\\[\\s*"                                 // constraints [optional]
+                    + "([\\+\\-])?\\s*"
+                    + "([A-Za-z0-9\\{\\}\\-\\.\\=\\*\\|:,/_\\s]+)"
+                + "\\])?\\s*$");
```

Comment:
Unnecessary escapes, and parentheses; naming parts adds clarity:
```java
            final String nameRE = "[A-Za-z0-9.-]+";
            final String fileRE = "[A-Za-z0-9./${}]+";
            Pattern p = Pattern.compile(
                  "^(" + nameRE + ")"                   // provider name
                + "\\s*"
                + "(" + fileRE + ")?"                   // configuration file [optional]
                + "\\s*"
                + "(\\["                                // constraints [optional]
                + "\\s*"
                    + "([+-])?"                         // action [optional]
                    + "[A-Za-z0-9{}.=*|:,/_\\s-]+"
                    + "\\])?"
                + "\\s*"
                + "$");
```
Should `fileRE` also accept `\`?
The string of constraints could be simplified to `[^\\]]+` (a sequence of one or more characters other than `]`).

Reply:
I added the backslash to the `fileRE`. I don't think we want to allow all characters other than `]`. We want to avoid as many potential user typos as possible by only allowing what is allowed to be used.

Changed the pattern to suggestion too.

Reply:
The backslash must be escaped in the Java literal and also in the regular expression. In Java you have to say `"\\\\"` to get a string with the two backslashes needed in the pattern.

Reply:
Fixed that.